### PR TITLE
Automatically test plxpr capture integration with `assert_valid`

### PR DESCRIPTION
**Pull request by albi3ro**
**Context:**

Following on from #5523, we need a good way of ensuring operations will integrate well with program capture.  

**Description of the Change:**

This PR adds a check to the `assert_valid` function to ensure capture is working as expected. 

It then starts to fix up some operations that were failing to be captured correctly.

**Benefits:**

More robust integration with pl capture.

**Possible Drawbacks:**

I'm realizing a lot more operators fail to be immediately captured than I initially assumed 😢 

**Related GitHub Issues:**

[sc-63310]

### CODE REVIEW

**Review by dwierichs — COMMENTED** [review]
LGTM, just a few tiny questions. Thanks for the patch @albi3ro!
I'm wondering: Can we have a systematic way to make sure that `assert_valid` is called on all PL objects it can be applied to? :thinking: 
Similar to the logic in the templates testing file, where PL is scraped for `Operator`s (and `SymbolicOp`s?) and `assert_valid` keeps track of which classes it saw? :thinking:

File: pennylane/ops/meta.py [hunk 1]
```diff
@@ -201,6 +201,11 @@ def circuit():
     num_params = 0
     grad_method = None
 
+    @classmethod
+    def _primitive_bind_call(cls, tag=None, measurement=None):
+        # TODO: make measurements dynamic
+        return cls._primitive.bind(measurement=measurement, tag=tag)
```
[the remark this answers is not in the document]
Comment (albi3ro) [reply pennylane/ops/meta.py:207]: Once measurements can also be captured, we can come back to this and make `measurement` positional.  At that point, we will also need to override the `def_impl` call.

File: pennylane/ops/op_math/controlled_ops.py [hunk 2]
```diff
@@ -209,6 +209,10 @@ def _flatten(self):
     def _unflatten(cls, data, metadata):
         return cls(metadata[0])
 
+    @classmethod
+    def _primitive_bind_call(cls, wires, id=None):
+        return cls._primitive.bind(*wires, n_wires=2)
```
[the remark this answers is not in the document]
Comment (albi3ro) [reply pennylane/ops/op_math/controlled_ops.py:214]: So we overrided `Controlled._primtiive.def_impl`. but that wouldn't override the primitive for all the child classes.  They would still have the default `Operator` behaviour.  

If/ when we create a `Operator._capture_def_impl` method and have the implementation method determined by inheritance, then we would have to once again fix it in these child classes.

[the remark this answers is not in the document]
Comment (dwierichs) [reply pennylane/ops/op_math/controlled_ops.py:214]: Ohhh, I see, thanks! :) I was assuming inheritance for some reason :facepalm:

**Review by dwierichs — APPROVED** [review]
Thanks for the explainers! :pray:

**Comment by albi3ro** [comment]
> LGTM, just a few tiny questions. Thanks for the patch @albi3ro! I'm wondering: Can we have a systematic way to make sure that `assert_valid` is called on all PL objects it can be applied to? 🤔 Similar to the logic in the templates testing file, where PL is scraped for `Operator`s (and `SymbolicOp`s?) and `assert_valid` keeps track of which classes it saw? 🤔

So by default we do automatically validate all operators except for templates.  See this file for the automatic collection: https://github.com/PennyLaneAI/pennylane/blob/master/tests/ops/functions/conftest.py

We opted templates out of this process due to the complexity of automatically generating a valid instance, but most templates have an `assert_valid` test associated with them.

**Comment by dwierichs** [comment]
Ah, brilliant! So the only thing one _could_ add is a cache that memorizes tested templates (from the template test files) and checks for untested templates (as per `assert_valid`). I can see this being very shaky though :upside_down_face:

File: pennylane/ops/meta.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (dwierichs) [reply pennylane/ops/meta.py:207]: Do we want to do this now? :upside_down_face:

File: pennylane/resource/first_quantization.py [hunk 3]
```diff
@@ -144,6 +144,22 @@ def _flatten(self):
...
+        eta,
+        omega=None,
+        error=0.0016,
+        charge=0,
+        br=7,
+        vectors=None,
+    ):
+        return cls._primitive.bind(
+            n, eta, omega=omega, error=error, charge=charge, br=br, vectors=vectors
+        )
```
Comment (dwierichs) [inline pennylane/resource/first_quantization.py:161]: I'm torn on this one: I read through it, trying to find out whether there is custom handling of any (kw)args. The number of arguments makes it hard to see this at once, so maybe the following is actually "more readable" in the end?
```suggestion
    # pylint: disable=arguments-differ
    @classmethod
    def _primitive_bind_call(cls, n, eta, **kwargs):
        return cls._primitive.bind(n, eta, **kwargs)
```

**Review by dwierichs — COMMENTED** [review]
Two new comments I had.

File: pennylane/ops/meta.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (albi3ro) [reply pennylane/ops/meta.py:207]: changing.

**Review by mudit2812 — APPROVED** [review]

**Pull request merged**